src/boot/efi/boot.c

```diff
@@ -2268,6 +2332,111 @@ static VOID config_write_entries_to_variable(Config *config) {
         (void) efivar_set_raw(LOADER_GUID, L"LoaderEntries", buffer, (UINT8 *) p - (UINT8 *) buffer, 0);
 }
 
+static EFI_STATUS secure_boot_enroll_from(Config *config, EFI_FILE_HANDLE root_dir, CHAR16 *keys_dir) {
+        static const UINT32 sb_vars_opts =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS |
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+                0;
+
+        static const struct {
+                const CHAR16 *name;
+                const CHAR16 *filename;
+                const EFI_GUID vendor;
+        } sb_vars[] = {
+                { L"db",  L"db.auth",  EFI_IMAGE_SECURITY_DATABASE_VARIABLE },
+                { L"KEK", L"KEK.auth", EFI_GLOBAL_VARIABLE },
+                { L"PK",  L"PK.auth",  EFI_GLOBAL_VARIABLE },
+        };
+
+        _cleanup_freepool_ CHAR8 *buffer = NULL;
+        UINTN size;
+        EFI_STATUS err = EFI_SUCCESS;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE dir = NULL;
+
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &dir, keys_dir, EFI_FILE_MODE_READ, 0ULL);
+        if (EFI_ERROR(err)) {
+                Print(L"Failed getting a handle on keys directory %s: %r\n", keys_dir, err);
+                return err;
+        }
+
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                err = file_read(dir, sb_vars[i].filename, 0, 0, &buffer, &size);
+                if (EFI_ERROR(err)) {
+                        Print(L"Failed reading %s file: %r\n", sb_vars[i].filename, err);
+                        return err;
+                }
+
+                err = efivar_set_raw(&sb_vars[i].vendor, sb_vars[i].name, buffer, size, sb_vars_opts);
+                if (EFI_ERROR(err)) {
+                        Print(L"Failed to write %s secure boot variable: %r\n", sb_vars[i].name, err);
+                        return err;
+                }
+        }
+
+        // if we loaded the signing keys successfully then we remove all the
+        // signing keys entries from  the menu as the system is now locked down.
+        config_remove_by_type(config, LOADER_SIGNATURE_KEYS);
+
+        Print(L"Finished loading signature keys from %s.\n", keys_dir);
+        uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+        return err;
+}
+
+static EFI_STATUS secure_boot_discover_keys(Config *config, EFI_FILE *root_dir) {
+        EFI_STATUS err = EFI_SUCCESS;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE keys_basedir = NULL;
+
+        if (!setup_mode_enabled())
+                return EFI_SUCCESS;
+
+        // if auto enrollement is activated (this is the default), we try to
+        // load keys straight from the /loader/keys/auto location. If it fails 
+        // (e.g. there is no such directory), we still proceed to add the
+        // entries to the menu (including the auto one)
```

Comment:
We use // for "fixme" style comments, and /* */ comments for explanatory comments. i.e. please switch from // to /* */ here.

See:
https://systemd.io/CODING_STYLE/